server/jobs/github_search.py

```diff
@@ -0,0 +1,196 @@
+    seen_repos = set()
+    treshold_date = arrow.utcnow().replace(weeks=(-1 * weeks_past))
+    for repo, details in repos.items():
+        api_url = details['repository']['url']
+        html_url = details['repository']['html_url']
+        if html_url in seen_repos:
+            continue
+        updated_time = make_github_req(api_url, access_token)['pushed_at']
+        time_update = arrow.get(updated_time)
+        if time_update > treshold_date:
+            owner = make_github_req(details['repository']['owner']['url'], access_token)
+            if not owner['name']:
+                owner['name'] = "@{}".format(owner['login'])
+            if not owner['email']:
+                owner['email'] = "?"
+
+            issue_repos[repo] = details
+            user_details = "{} ({})".format(owner['name'], owner['email'])
+            logger.info("{}: {} - {}".format(user_details, html_url,
+                                             time_update.humanize()))
+        seen_repos.add(details['repository']['html_url'])
+    return issue_repos
+
+def get_online_repos(source, logger, language, access_token,
+                     keyword='def '):
+    longest_lines = get_longest_defs(source, keyword)
+    if not longest_lines:
+        logger.warning("'{}' not found in source file: {}".format(keyword, source))
+        return
+    else:
+        logger.info("Found {} candidate lines".format(len(longest_lines)))
+    source = []
+
+    for i in range(min(2, len(longest_lines))):
+        logger.info("Fetching query {}".format(i))
+        res = get_all_results(longest_lines[i], language, logger, access_token)
+        source.extend(res)
+        time.sleep(9)
+
+    repos = OrderedDict()
+    for repo in source:
+        if not repo:
+            continue
+        else:
+            repos[repo['html_url']] = repo
+    return repos
+
+@jobs.background_job
+def search_similar_repos(access_token=None, assignment_id=None,
+                         language='python', template_name=None,
+                         keyword='def ', weeks_past=12):
+    logger = jobs.get_job_logger()
+    logger.info('Starting Github Search...')
+
+    assign = Assignment.query.filter_by(id=assignment_id).one_or_none()
+    if not assign:
+        logger.info("Could not find assignment")
+        return
+    if not assign.files:
+        logger.info("Upload template files for this assignment to search.")
+        return
+
+    possible_file_names = list(assign.files.keys())
+    if template_name not in possible_file_names:
+        logger.info("{} is not in {}".format(template_name, possible_file_names))
+
+    source_file = assign.files[template_name]
+    repos = get_online_repos(source_file, logger, language, access_token,
+                             keyword=keyword)
+    if not repos:
+        logger.warning("No repos found. Try a different keyword?")
+        return
+    recent_repos = list_recent_repos(repos, logger, access_token, weeks_past)
```

Comment:
Good point. I left it in because I plan on using it for the future improvement (filing issues). There are "side-effects" in that things get logged (what I care about for now)